src/overlays/actors/ovl_En_Kitan/z_en_kitan.c

```diff
@@ -27,50 +30,364 @@ ActorInit En_Kitan_InitVars = {
+        TOUCH_NONE | TOUCH_SFX_NORMAL,
+        BUMP_ON,
+        OCELEM_ON,
+    },
     { 20, 40, 0, { 0, 0, 0 } },
 };
 
-#endif
+void EnKitan_Init(Actor* thisx, PlayState* play) {
+    EnKitan* this = THIS;
+    s32 pad;
+
+    Actor_SetScale(&this->actor, 0.0f);
+    this->actionFunc = EnKitan_WaitToAppear;
+
+    ActorShape_Init(&this->actor.shape, 0.0f, ActorShadow_DrawCircle, 12.0f);
+    SkelAnime_InitFlex(play, &this->skelAnime, &gKeatonSkel, &gKeatonIdleAnim, this->jointTable, this->morphTable,
+                       KEATON_LIMB_MAX);
+    Animation_PlayLoop(&this->skelAnime, &gKeatonIdleAnim);
+
+    Collider_InitAndSetCylinder(play, &this->collider, &this->actor, &sCylinderInit);
+    this->actor.colChkInfo.mass = MASS_IMMOVABLE;
+    Collider_UpdateCylinder(&this->actor, &this->collider);
+
+    this->actor.velocity.y = -9.0f;
+    this->actor.terminalVelocity = -9.0f;
+    this->actor.gravity = -1.0f;
+
+    if (Player_GetMask(play) != PLAYER_MASK_KEATON ||
+        Flags_GetCollectible(play, ENKITAN_GET_COLLECT_FLAG(&this->actor))) {
+        Actor_Kill(&this->actor);
+        return;
+    }
+
+    this->timer = 120;
+    this->actor.flags &= ~ACTOR_FLAG_TARGETABLE;
+}
+
+void EnKitan_Destroy(Actor* thisx, PlayState* play) {
+    EnKitan* this = THIS;
+
+    Collider_DestroyCylinder(play, &this->collider);
+}
+
+void EnKitan_SpawnEffects(EnKitan* this, PlayState* play, s32 numEffects) {
+    static Color_RGBA8 sEffPrimColor = { 255, 255, 255, 255 };
+    static Color_RGBA8 sEffEnvColor = { 255, 255, 200, 255 };
+    s32 i;
+    Vec3f accel;
+    Vec3f vel;
+    Vec3f pos;
+
+    pos.x = this->actor.world.pos.x;
+    pos.y = this->actor.world.pos.y;
+    pos.z = this->actor.world.pos.z;
+
+    for (i = 0; i < numEffects; i++) {
+        vel.x = Rand_CenteredFloat(10.0f);
+        vel.y = Rand_ZeroFloat(6.0f);
+        vel.z = Rand_CenteredFloat(10.0f);
+        accel.x = -vel.x * 0.05f;
+        accel.y = 0.1f;
+        accel.z = -vel.x * 0.05f;
+        func_800B0F18(play, &pos, &vel, &accel, &sEffPrimColor, &sEffEnvColor, 400, 20, 20);
+    }
+}
+
+s32 EnKitan_CanTalk(EnKitan* this, PlayState* play) {
+    Player* player = GET_PLAYER(play);
+
+    if (player->stateFlags1 & PLAYER_STATE1_800000) {
+        return false;
+    }
+    if (Player_IsFacingActor(&this->actor, 0x3000, play) && Actor_IsFacingPlayer(&this->actor, 0x3000) &&
+        this->actor.xzDistToPlayer < 120.0f) {
+        return true;
+    }
+    return false;
+}
+
+u16 EnKitan_GetQuestionMessageId(EnKitan* this) {
+    s32 i = 0;
+
+    //! @bug This loop was likely meant to have a max iteration cap of 1000, but the condition is messed up so it will
+    //! loop forever until a question that was not yet asked is found. This is fine as the number of possible questions
+    //! is larger than the number of questions the player is required to answer, assuming the random number generator
+    //! can output the full range of values which is the case.
```

Comment:
Hmm, I'm not 100% sure about this. Are you sure the *condition* is messed up? If I'm reading this right, it seems to me that this could've worked to get out of the loop after 1000 iterations:
```c
        if (i > 1000) {
            break;
        }
```

However, it would result in nothing being returned. I'm not really sure what they intended to happen after 1000 iterations assuming that they wanted to break out. So while you currently get stuck in here forever until you find a valid question, that at least means that this function will return properly every time. Maybe they originally wanted this to break out after 1000 iterations, realized they didn't really have a good return value in that case, and just decided to make it loop forever?

Reply:
IDO can sometime optimize out code if it sees there's no way to reach it. Testing on decomp.me seems like you can put a return after the loop and it still matches
https://decomp.me/scratch/AB3Nq

Reply:
@tom-overton there's no situation where it would be acceptable to allow the loop to run forever and lock up the game.

I feel it's proper to write the function with a final return statement, but with a comment stating that the return value is unknown, as you can put pretty much anything after the while loop and it will match.